pkgs/dart_model/lib/src/scopes.dart

```diff
@@ -118,6 +123,47 @@ enum Scope {
 class _ScopeData {
   final Scope type;
   final JsonBufferBuilder? buffer;
+  final MacroScope? macro;
+
+  _ScopeData._(this.type, this.buffer, [this.macro]);
+}
+
+/// Data attached to [Scope.macro] environments to make resolved queries
+/// publicly accessible.
+class MacroScope {
+  Model? _accumulatedModel;
+
+  /// A cached [StaticTypeSystem] instance bound to the resolved type hierarchy
+  /// from the accumulated model.
+  ///
+  /// Since new models added through [addModel] only contribute new aspects of
+  /// the type system without invalidating earlier incomplete views, all queries
+  /// made through earlier type system instances continue to be valid.
+  StaticTypeSystem? _typeSystem;
 
-  _ScopeData._(this.type, this.buffer);
+  MacroScope._();
+
+  /// Returns a type system resolving subtype queries for types known to the
+  /// current Dart model.
+  StaticTypeSystem get typeSystem {
+    return _typeSystem ??= switch (_accumulatedModel) {
+      // TODO: Refactor type system so that we can return an empty type system
+      // here that doesn't know anything?
+      null => throw StateError('Cannot obtain type system without a model'),
+      var model => StaticTypeSystem(model.types),
+    };
+  }
+
+  void addModel(Model model) {
+    // TODO: Integrate new model into accumulated model instead of replacing it
```

Comment:
I suggest stating the intent, "merge `model` into the scope's accumulated model", in a doc comment, then adding the TODO right next to that clarifying that that is not what it actually does today :)